### PR TITLE
wasm: return more structured errors, safe interrupt handling

### LINKED ISSUE #3225: errors when using the 'wasm' target are bleak
👉 https://github.com/open-policy-agent/opa/pull/3211/files#r588449268

```
(sr/wasm/context)torin:~/go/src/github.com/open-policy-agent/opa$ ./opa_linux_amd64 test --timeout=1s ~/x.rego -v
data.x.test_xxx: ERROR (1.0009364s)
  eval_cancel_error: numbers.range: timed out before generating all numbers in range
--------------------------------------------------------------------------------
ERROR: 1/1
(sr/wasm/context)torin:~/go/src/github.com/open-policy-agent/opa$ ./opa_linux_amd64 test --timeout=1s ~/x.rego -t wasm
data.x.test_xxx: ERROR (1.0010134s)
  cancelled
--------------------------------------------------------------------------------
ERROR: 1/1
```

It would be nice if these were more alike!

### PR DESCRIPTION
**Pull request by srenatus**
We'll now return properly structured objects, with code+message:

    $ opa test -t wasm --timeout 1s .  -v
    data.p.test_that_takes_long: ERROR (1.004488054s)
      cancelled: interrupted at eval/g0.data.p.test_that_takes_long/opa_numbers_range/qadd_one/mpd_del/free/opa_free
    --------------------------------------------------------------------------------
    ERROR: 1/1
    $ opa test -t wasm --timeout 1s .  -v --format=json
    [
      {
        "location": {
          "file": "t.rego",
          "row": 3,
          "col": 1
        },
        "package": "data.p",
        "name": "test_that_takes_long",
        "error": {
          "code": "cancelled",
          "message": "interrupted at eval/g0.data.p.test_that_takes_long/opa_numbers_range/opa_array_append"
        },
        "duration": 1003761215
      }
    ]

The interrupt stacktrace is an attempt to provide some useful information,
similar to the topdown case; but the scene naturally looks different in wasm.

Fixes #3225.

----

The three different types of cancellation involved in the wasm code path now all get their error _messages_ bubbled up as well:

- plain interrupted wasm eval (see above)
- host builtin implementations cancelled using the topdown.Cancel machinery: `cancelled: net.cidr_expand: timed out before generating all IP addresses`
- host builtin implementations cancelled using context.Context's deadline: `http.send: timed out (context deadline exceeded)`

---------
Also small improvements to topdown error matcher implementations -- they now work with wrapped errors.

------

Bumping golang runtime version to avoid a runtime explosion eventually happening when running the interrupt stuff in a loop. In some back and forth with the fine folks maintaining wasmtime-go, it came up that the previous handling of the interrupt, using the store in a separate go routine, wasn't safe, so this has been changed here, too.

### CODE REVIEW

**Review by tsandall — COMMENTED** [review]
LGTM! Nice.

**Review by tsandall — APPROVED** [review]

File: topdown/errors.go
Comment (srenatus) [reply topdown/errors.go]: The short answer is yes. The longer one is you got me thinking about this some more and I think we ought to include all fields in this implementation, so it's not confusing to users. Done that now.

**Comment by srenatus** [comment]
🚧 I'm seeing "runtime pc" errors every now and then, I'll have to get to the bottom of this before we merge it.

**Comment by srenatus** [comment]
I don't yet understand how _this PR_ changes things, but revisiting an older wasmtime-go issue shows that it's still not fully done, unfortunately: https://github.com/bytecodealliance/wasmtime-go/issues/60#issuecomment-823252710

**Comment by srenatus** [comment]
Interestingly, this time https://github.com/open-policy-agent/opa/commit/db046c39feb6fc61ac7549f3f10f27f3b5c0cf51 doesn't make a difference. Probably because interrupt traps aren't completely different in handling from the unreachable traps.

**Comment by srenatus** [comment]
OK so the footgun here was printing the error:
```go
if testing.Verbose() {
	t.Logf("err: %s, %T", err.Error(), err)
}
```
that'll sometimes trigger the SIGILL in wasmtime/cgo interaction... So, while this is annoying, and I'll share an MWE with our wasmtime-go friends, it doesn't have to block this PR 😳

**Comment by srenatus** [comment]
🚧 still happens, don't merge 🚧

**Comment by srenatus** [comment]
Problem went away with upgrading to Golang 1.16. This means we shouldn't merge this until we've updated the runtime version we use to 1.16. We can still keep compat using go.mod. Adding a commit for this.

**Comment by srenatus** [comment]
⚠️ Still unresolved. With Go 1.16, I get spurious SIGILLs still:

```
$ ./opa_darwin_amd64 eval -t wasm --format=pretty '{ x: y | x := "a"; y := [3, 4][_] }'
SIGILL: illegal instruction
PC=0x2f3cd364 m=0 sigcode=1
instruction bytes: 0xf 0xb 0xf 0xb 0x55 0x48 0x89 0xe5 0x4c 0x8b 0x17 0x4d 0x8b 0x12 0x49 0x39

goroutine 0 [idle]:
runtime: unknown pc 0x2f3cd364
stack: frame={sp:0x7ffeefbff200, fp:0x0} stack=[0x7ffeefb80738,0x7ffeefbff7a0)
00007ffeefbff100:  00007ffeefbff140  00000000048b101a
00007ffeefbff110:  0000000000000007  0000000008604a60
00007ffeefbff120:  0000000008605520  0000000030060cc0
00007ffeefbff130:  00007ffeefbff180  0000000030060cc0
00007ffeefbff140:  00007ffeefbff170  0000000004c1284f
00007ffeefbff150:  0000000008605a50  000000000001dc00
00007ffeefbff160:  0000000000000024  000000000001dbd0
00007ffeefbff170:  00007ffeefbff190  00000000082e0018
00007ffeefbff180:  000000000001e0a8  000000000001e0a8
00007ffeefbff190:  00007ffeefbff1f0  000000002f3b039e
00007ffeefbff1a0:  000000000000dc6d  000000000000dc92
00007ffeefbff1b0:  000000000001dc30  000000000001dbc0
00007ffeefbff1c0:  0000000030060cc0  000000000001e040
00007ffeefbff1d0:  000000000001dff4  000000000001de8c
00007ffeefbff1e0:  000000000001def4  0000000000000000
00007ffeefbff1f0:  00007ffeefbff240  000000002f3cd364
00007ffeefbff200: <000000000001de24  0000000008604220
00007ffeefbff210:  0000000006b109b0  00007ffeefbff270
00007ffeefbff220:  00007ffeefbff510  0000000000000000
00007ffeefbff230:  0000000004b8dc00  0000000000000028
00007ffeefbff240:  00007ffeefbff260  000000002f3cd550
00007ffeefbff250:  00007ffeefbff390  00007fff201cb7aa
00007ffeefbff260:  00007ffeefbff340  0000000004f83eec
00007ffeefbff270:  0000000000000000  58356ce67e91db79
00007ffeefbff280:  58356ce67e91da49  00007ffeefbff390
00007ffeefbff290:  00007ffeefbff270  00007ffeefbff510
00007ffeefbff2a0:  0000000004b8dc00  5835131895d616e7
00007ffeefbff2b0:  0000000000000000  0000037f00001fa0
00007ffeefbff2c0:  0000000000000000  0000000000003c00
00007ffeefbff2d0:  0000000000000001  0000000000000001
00007ffeefbff2e0:  00000000068d0a00  00000000068d4600
00007ffeefbff2f0:  00007ffeefbff390  00007fff201c8b79
runtime: unknown pc 0x2f3cd364
stack: frame={sp:0x7ffeefbff200, fp:0x0} stack=[0x7ffeefb80738,0x7ffeefbff7a0)
00007ffeefbff100:  00007ffeefbff140  00000000048b101a
00007ffeefbff110:  0000000000000007  0000000008604a60
00007ffeefbff120:  0000000008605520  0000000030060cc0
00007ffeefbff130:  00007ffeefbff180  0000000030060cc0
00007ffeefbff140:  00007ffeefbff170  0000000004c1284f
00007ffeefbff150:  0000000008605a50  000000000001dc00
00007ffeefbff160:  0000000000000024  000000000001dbd0
00007ffeefbff170:  00007ffeefbff190  00000000082e0018
00007ffeefbff180:  000000000001e0a8  000000000001e0a8
00007ffeefbff190:  00007ffeefbff1f0  000000002f3b039e
00007ffeefbff1a0:  000000000000dc6d  000000000000dc92
00007ffeefbff1b0:  000000000001dc30  000000000001dbc0
00007ffeefbff1c0:  0000000030060cc0  000000000001e040
00007ffeefbff1d0:  000000000001dff4  000000000001de8c
00007ffeefbff1e0:  000000000001def4  0000000000000000
00007ffeefbff1f0:  00007ffeefbff240  000000002f3cd364
00007ffeefbff200: <000000000001de24  0000000008604220
00007ffeefbff210:  0000000006b109b0  00007ffeefbff270
00007ffeefbff220:  00007ffeefbff510  0000000000000000
00007ffeefbff230:  0000000004b8dc00  0000000000000028
00007ffeefbff240:  00007ffeefbff260  000000002f3cd550
00007ffeefbff250:  00007ffeefbff390  00007fff201cb7aa
00007ffeefbff260:  00007ffeefbff340  0000000004f83eec
00007ffeefbff270:  0000000000000000  58356ce67e91db79
00007ffeefbff280:  58356ce67e91da49  00007ffeefbff390
00007ffeefbff290:  00007ffeefbff270  00007ffeefbff510
00007ffeefbff2a0:  0000000004b8dc00  5835131895d616e7
00007ffeefbff2b0:  0000000000000000  0000037f00001fa0
00007ffeefbff2c0:  0000000000000000  0000000000003c00
00007ffeefbff2d0:  0000000000000001  0000000000000001
00007ffeefbff2e0:  00000000068d0a00  00000000068d4600
00007ffeefbff2f0:  00007ffeefbff390  00007fff201c8b79

goroutine 1 [syscall]:
runtime.cgocall(0x4885ec0, 0xc0004fadd8, 0xc0003362b0)
        /nix/store/2dgrrcr9z6571yygbzwcds3ss1vid4sk-go-1.16/share/go/src/runtime/cgocall.go:154 +0x5b fp=0xc0004fada8 sp=0xc0004fad70 pc=0x400769b
github.com/bytecodealliance/wasmtime-go._Cfunc_go_wasmtime_func_call(0x6b0ee10, 0xc0003362a0, 0xc0003362b0, 0xc0002b3a28, 0x0)
        _cgo_gotypes.go:715 +0x49 fp=0xc0004fadd8 sp=0xc0004fada8 pc=0x4553109
github.com/bytecodealliance/wasmtime-go.(*Func).Call.func3.1(0xc0002feec0, 0xc0003362a0, 0xc0003362b0, 0xc0002b3a28, 0x40ce11ec6b340201)
        /Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/bytecodealliance/wasmtime-go/func.go:505 +0x117 fp=0xc0004fae18 sp=0xc0004fadd8 pc=0x4563817
github.com/bytecodealliance/wasmtime-go.(*Func).Call.func3(0xc0002b3a28)
        /Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/bytecodealliance/wasmtime-go/func.go:505 +0x4f fp=0xc0004fae58 sp=0xc0004fae18 pc=0x456388f
github.com/bytecodealliance/wasmtime-go.enterWasm(0xc0002a8120, 0xc0004faf80, 0x0)
        /Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/bytecodealliance/wasmtime-go/func.go:589 +0xf0 fp=0xc0004faeb8 sp=0xc0004fae58 pc=0x455baf0
github.com/bytecodealliance/wasmtime-go.(*Func).Call(0xc0002feec0, 0xc000336260, 0x1, 0x1, 0xc000336270, 0xc0002feec0, 0xc0004fb020, 0x455d2cf)
        /Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/bytecodealliance/wasmtime-go/func.go:500 +0x27d fp=0xc0004fafb8 sp=0xc0004faeb8 pc=0x455b2bd
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callOrCancel.func2(0xc0007086c0, 0xc0002feec0, 0xc000336260, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/internal/wasm/vm.go:633 +0xb7 fp=0xc0004fb030 sp=0xc0004fafb8 pc=0x46032b7
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callOrCancel(0x57ae5d0, 0xc00012e008, 0xc00029a100, 0x568bd44, 0x4, 0xc0004fb164, 0x1, 0x1, 0x8a58f25, 0x5eb6f40, ...)
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/internal/wasm/vm.go:635 +0x1d6 fp=0xc0004fb100 sp=0xc0004fb030 pc=0x4601ab6
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callVoid(...)
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/internal/wasm/vm.go:584
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.newVM.func1(0x57ae5d0, 0xc00012e008, 0x1dd8c, 0x57ae7c8, 0xc0003302a0)
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/internal/wasm/vm.go:105 +0x89 fp=0xc0004fb178 sp=0xc0004fb100 pc=0x4602229
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.(*VM).Eval(0xc00029a100, 0x57ae5d0, 0xc00012e008, 0x0, 0x0, 0x57b4768, 0xc0002b4ac0, 0x0, 0x0, 0x0, ...)
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/internal/wasm/vm.go:276 +0x34f fp=0xc0004fb220 sp=0xc0004fb178 pc=0x45ffb4f
github.com/open-policy-agent/opa/internal/wasm/sdk/opa.(*OPA).Eval(0xc0000b6c40, 0x57ae5d0, 0xc00012e008, 0x0, 0x0, 0x57b4768, 0xc0002b4ac0, 0x0, 0x0, 0x0, ...)
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/opa/opa.go:185 +0x1aa fp=0xc0004fb308 sp=0xc0004fb220 pc=0x4604cca
github.com/open-policy-agent/opa/internal/rego/opa.(*OPA).Eval(0xc000010038, 0x57ae5d0, 0xc00012e008, 0x0, 0x57b4768, 0xc0002b4ac0, 0x0, 0x0, 0x0, 0x6db8000, ...)
        /Users/stephan/Sources/open-policy-agent/opa/internal/rego/opa/opa.go:55 +0x109 fp=0xc0004fb3c0 sp=0xc0004fb308 pc=0x4607c29
github.com/open-policy-agent/opa/rego.(*Rego).evalWasm(0xc000308a00, 0x57ae5d0, 0xc00012e008, 0xc0002ba140, 0x0, 0x203000, 0x203000, 0x203000, 0x82a1cb8)
        /Users/stephan/Sources/open-policy-agent/opa/rego/rego.go:1913 +0x106 fp=0xc0004fb4a0 sp=0xc0004fb3c0 pc=0x4612786
github.com/open-policy-agent/opa/rego.(*Rego).eval(0xc000308a00, 0x57ae5d0, 0xc00012e008, 0xc0002ba140, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/stephan/Sources/open-policy-agent/opa/rego/rego.go:1840 +0x81f fp=0xc0004fb6d8 sp=0xc0004fb4a0 pc=0x46125df
github.com/open-policy-agent/opa/rego.PreparedEvalQuery.Eval(0xc000308a00, 0xc000292f10, 0x57ae5d0, 0xc00012e008, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/stephan/Sources/open-policy-agent/opa/rego/rego.go:353 +0x19b fp=0xc0004fb768 sp=0xc0004fb6d8 pc=0x46090fb
github.com/open-policy-agent/opa/cmd.eval(0xc0002a29c0, 0x1, 0x4, 0x0, 0xc0002927d0, 0x1, 0x1, 0x5eefd00, 0x0, 0x0, ...)
        /Users/stephan/Sources/open-policy-agent/opa/cmd/eval.go:291 +0xaa5 fp=0xc0004fba58 sp=0xc0004fb768 pc=0x4874fe5
github.com/open-policy-agent/opa/cmd.init.4.func2(0xc0002e3180, 0xc0002a29c0, 0x1, 0x4)
        /Users/stephan/Sources/open-policy-agent/opa/cmd/eval.go:222 +0xcf fp=0xc0004fbd88 sp=0xc0004fba58 pc=0x4882e4f
github.com/spf13/cobra.(*Command).execute(0xc0002e3180, 0xc0002a2980, 0x4, 0x4, 0xc0002e3180, 0xc0002a2980)
        /Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:766 +0x2c2 fp=0xc0004fbe48 sp=0xc0004fbd88 pc=0x4646282
github.com/spf13/cobra.(*Command).ExecuteC(0x5ea8460, 0xc000000180, 0xc00021ff78, 0x400a265)
        /Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:852 +0x2fe fp=0xc0004fbf28 sp=0xc0004fbe48 pc=0x4646e9e
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:800
main.main()
        /Users/stephan/Sources/open-policy-agent/opa/main.go:15 +0x31 fp=0xc0004fbf88 sp=0xc0004fbf28 pc=0x4884931
runtime.main()
        /nix/store/2dgrrcr9z6571yygbzwcds3ss1vid4sk-go-1.16/share/go/src/runtime/proc.go:225 +0x256 fp=0xc0004fbfe0 sp=0xc0004fbf88 pc=0x403e8b6
runtime.goexit()
        /nix/store/2dgrrcr9z6571yygbzwcds3ss1vid4sk-go-1.16/share/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc0004fbfe8 sp=0xc0004fbfe0 pc=0x4076261

goroutine 27 [select]:
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callOrCancel.func1(0x57ae5d0, 0xc00012e008, 0xc00029a100, 0xc0007086c0, 0xc000708720)
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/internal/wasm/vm.go:605 +0x87
created by github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callOrCancel
        /Users/stephan/Sources/open-policy-agent/opa/internal/wasm/sdk/internal/wasm/vm.go:604 +0x16d

rax    0x0
rbx    0x1de8c
rcx    0x1db90
rdx    0x4075f2d
rdi    0x7ffeefbff118
rsi    0x1dc30
rbp    0x7ffeefbff240
rsp    0x7ffeefbff200
r8     0x68d0a00
r9     0x6b00000
r10    0x1
r11    0x6b109c0
r12    0x30060cc0
r13    0x1e040
r14    0x1dff4
r15    0x1def4
rip    0x2f3cd364
rflags 0x10206
cs     0x2b
fs     0x0
gs     0x0
$ .//opa_darwin_amd64 eval -t wasm --format=pretty '{ x: y | x := "a"; y := [3, 4][_] }'
1 error occurred: internal_error: <query>:1:1: object insert conflict
$ ./opa_darwin_amd64 eval -t wasm --format=pretty '{ x: y | x := "a"; y := [3, 4][_] }'
1 error occurred: internal_error: <query>:1:1: object insert conflict
```

**Comment by srenatus** [comment]
Bumped wasmtime-go. Looks like it's all resolved 🎉

**Review by tsandall — APPROVED** [review]

**Pull request merged**
